### ISSUE TITLE
optimizations for generated target code

### ISSUE THREAD

**Issue by rolph-recto** [apl-cornell/viaduct#27]
the generated target code is rather verbose right now. for example, here is the spec program and output for the "dangerous" millionaires game:

spec:
```
/* the classic MPC problem */
a : {A};
c : {C};
c_richer : {(A|C)-> & (A&C)<-};
c_richer := declassify(a <= c, {(A|C)-> & (A&C)<-});

/* the dangerous version: if you have less money, you lose all of it */
if (c_richer) {
    a := 0;
} else {
    c := 0;
}
```

generated target code:
```
MPC({alice,chuck})_1 {
    c_1 <- recv chuck;
    a_1 <- recv alice;
    downgrade_1_1 := (a_1 <= c_1);
    send downgrade_1_1 to alice;
    send downgrade_1_1 to chuck;
}

alice {
    a : {⊥<-};
    c_richer : {⊥<-};
    send a to MPC({alice,chuck})_1;
    downgrade_1_2 <- recv MPC({alice,chuck})_1;
    assgn_1_1 := downgrade_1_2;
    c_richer := assgn_1_1;
    c_richer_1 := c_richer;
    guard_1_1 := c_richer_1;
    guard_1_4 := guard_1_1;
    if (guard_1_4) {
        assgn_2_1 := 0;
        a := assgn_2_1;
        a <- recv chuck;
    } else {
        assgn_3_1 := 0;
        send assgn_3_1 to chuck;
    }
}

chuck {
    c : {⊥<-};
    c_richer : {⊥<-};
    send c to MPC({alice,chuck})_1;
    downgrade_1_3 <- recv MPC({alice,chuck})_1;
    assgn_1_2 := downgrade_1_3;
    c_richer := assgn_1_2;
    c_richer_2 := c_richer;
    guard_1_2 := c_richer_2;
    guard_1_3 := guard_1_2;
    if (guard_1_3) {
        assgn_2_2 := 0;
        send assgn_2_2 to alice;
    } else {
        c <- recv alice;
        assgn_3_2 := 0;
        c := assgn_3_2;
    }
}
```

as you can see, there are a lot of unnecessary variables. we can run copy propagation, dead code elimination, and constant propagation to remove most of these and make the target a lot readable.

**Comment by rolph-recto** [comment]
copy prop implemented in https://github.com/apl-cornell/viaduct/commit/347d4cac7362e8f16c3bd132486accea58f27348

constant prop not yet implemented

**Comment by rolph-recto** [comment]
constant prop implemented in e3ccfd719801c6924b899335289b35280e6d2e6f
